File: services/nextjs/scripts/download-icons.py
import os
import urllib.request
import json
from concurrent.futures import ThreadPoolExecutor, as_completed
import time

ICONS_DIR = os.path.join(os.path.dirname(__file__), '../public/assets/icons')
MAPPING_URL = 'https://prices.runescape.wiki/api/v1/osrs/mapping'
USER_AGENT = 'OSRSMarketTracker/1.0 (contact: samirkatakamsetty@gmail.com; developer project)'

os.makedirs(ICONS_DIR, exist_ok=True)

def download_icon(item):
    item_id = item.get('id')
    if not item_id:
        return 'skipped'

    dest_path = os.path.join(ICONS_DIR, f"{item_id}.png")
    if os.path.exists(dest_path):
        return 'skipped'

    # Download directly from OSRSBox CDN
    cdn_url = f"https://www.osrsbox.com/osrsbox-db/items-icons/{item_id}.png"

    attempts = 3
    for attempt in range(1, attempts + 1):
        try:
            req = urllib.request.Request(
                cdn_url,
                headers={'User-Agent': USER_AGENT}
            )
            with urllib.request.urlopen(req, timeout=10) as response:
                with open(dest_path, 'wb') as f:
                    f.write(response.read())
            return 'downloaded'
        except Exception as e:
            if attempt == attempts:
                print(f"Failed to download icon for item {item.get('name')} ({item_id}) after {attempts} attempts: {e}")
                return 'failed'
            time.sleep(1.5 ** attempt)
    return 'failed'

def main():
    print("Fetching item mapping from OSRS Wiki...")
    req = urllib.request.Request(
        MAPPING_URL,
        headers={'User-Agent': USER_AGENT}
    )
    
    try:
        with urllib.request.urlopen(req, timeout=15) as response:
            items = json.loads(response.read().decode('utf-8'))
    except Exception as e:
        print(f"Fatal: Failed to fetch mapping: {e}")
        return

    print(f"Found {len(items)} items in mapping. Starting parallel download pool (45 threads) from OSRSBox CDN...")

    downloaded = 0
    skipped = 0
    failed = 0
    total = len(items)

    start_time = time.time()
    
    with ThreadPoolExecutor(max_workers=45) as executor:
        # Submit tasks
        future_to_item = {executor.submit(download_icon, item): item for item in items}
        
        processed = 0
        for future in as_completed(future_to_item):
            processed += 1
            res = future.result()
            if res == 'downloaded':
                downloaded += 1
            elif res == 'skipped':
                skipped += 1
            elif res == 'failed':
                failed += 1

            if processed % 100 == 0 or processed == total:
                print(f"Progress: {processed}/{total} processed (Downloaded: {downloaded}, Skipped: {skipped}, Failed: {failed})")

    duration = time.time() - start_time
    print(f"Download finished in {duration:.1f}s.")
    print(f"Stats: Total Downloaded = {downloaded}, Skipped = {skipped}, Failed = {failed}")

if __name__ == '__main__':
    main()

File: services/nextjs/scripts/download-icons.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const ICONS_DIR = path.join(__dirname, '../public/assets/icons');
const MAPPING_URL = 'https://prices.runescape.wiki/api/v1/osrs/mapping';
const USER_AGENT = 'OSRSMarketTracker/1.0 (contact: samirkatakamsetty@gmail.com; developer project)';

// Ensure directory exists
if (!fs.existsSync(ICONS_DIR)) {
  fs.mkdirSync(ICONS_DIR, { recursive: true });
}

async function downloadIcon(item, attempt = 1) {
  if (!item.icon) return;
  
  const destPath = path.join(ICONS_DIR, `${item.id}.png`);
  if (fs.existsSync(destPath)) {
    // Skip if already exists
    return { id: item.id, status: 'skipped' };
  }

  // Construct Special:FilePath URL
  const wikiUrl = `https://oldschool.runescape.wiki/w/Special:FilePath/${encodeURIComponent(item.icon)}`;

  try {
    const res = await fetch(wikiUrl, {
      headers: {
        'User-Agent': USER_AGENT
      }
    });

    if (!res.ok) {
      throw new Error(`HTTP error! status: ${res.status}`);
    }

    const buffer = await res.arrayBuffer();
    await fs.promises.writeFile(destPath, Buffer.from(buffer));
    return { id: item.id, status: 'downloaded' };
  } catch (err) {
    if (attempt < 3) {
      const delay = Math.pow(2, attempt) * 1000;
      // Wait and retry
      await new Promise(resolve => setTimeout(resolve, delay));
      return downloadIcon(item, attempt + 1);
    } else {
      console.error(`Failed to download icon for item ${item.name} (${item.id}) after 3 attempts:`, err.message);
      return { id: item.id, status: 'failed', error: err.message };
    }
  }
}

async function runPool(items, limit) {
  let active = 0;
  let index = 0;
  let completedCount = 0;
  const total = items.length;
  let downloadedCount = 0;
  let skippedCount = 0;
  let failedCount = 0;

  return new Promise((resolve) => {
    function next() {
      if (index >= total && active === 0) {
        resolve({ downloadedCount, skippedCount, failedCount });
        return;
      }

      while (active < limit && index < total) {
        const item = items[index++];
        active++;
        
        downloadIcon(item)
          .then((res) => {
            if (res) {
              if (res.status === 'downloaded') downloadedCount++;
              if (res.status === 'skipped') skippedCount++;
              if (res.status === 'failed') failedCount++;
            }
          })
          .catch((err) => {
            console.error('Unexpected pool error:', err);
            failedCount++;
          })
          .finally(() => {
            active--;
            completedCount++;
            if (completedCount % 100 === 0 || completedCount === total) {
              console.log(`Progress: ${completedCount}/${total} items processed (Downloaded: ${downloadedCount}, Skipped: ${skippedCount}, Failed: ${failedCount})`);
            }
            next();
          });
      }
    }
    next();
  });
}

async function main() {
  console.log('Fetching item mapping from OSRS Wiki...');
  try {
    const res = await fetch(MAPPING_URL, {
      headers: {
        'User-Agent': USER_AGENT
      }
    });
    if (!res.ok) {
      throw new Error(`Failed to fetch mapping: ${res.statusText}`);
    }
    const items = await res.json();
    console.log(`Found ${items.length} items in mapping. Starting download pool...`);

    const start = Date.now();
    const stats = await runPool(items, 15);
    const duration = ((Date.now() - start) / 1000).toFixed(1);

    console.log(`Download finished in ${duration}s.`);
    console.log(`Stats: Total Downloaded = ${stats.downloadedCount}, Skipped = ${stats.skippedCount}, Failed = ${stats.failedCount}`);
  } catch (err) {
    console.error('Fatal error in downloader:', err);
  }
}

main();
